### DISCUSSION TITLE
How to read data from L1

### DISCUSSION THREAD

**Discussion by girishbarca** [ethereum-optimism/optimistic-specs#64] [Q&A]
I'm trying to better understand how to read l1 contract variables/call view functions from l1 from an optimism l2 contract. Wondering if anyone had an idea as to how that could be done?
Thanks!

**Comment by norswap** [comment]
Citing an internal conversation we had on this topic:

> Ori: We could have a precompile that receives a static function call, and returns "immediately after block A, the result of this call is B". That is something whose accuracy verifiers can verify.
>
> norswap: ^ what Ori said
>
> We already expect validators to run a full node (because they need to access logs). Or if they don't because e.g. they query infura, then they can query it for state access too.
>
> In the fraud proof, it means we need to add a new "instruction" to the that reads from L1, and we need to be able to prove L1 storage slots to seed the oracle. Seems easy enough (it's just a slight variant of doing the same thing for L2 storage).

We haven't decided yet if that's something we want to do or when, but we're definitely thinking about it.
